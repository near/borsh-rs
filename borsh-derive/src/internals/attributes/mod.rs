use syn::{Attribute, Path};

pub mod field;
pub mod item;
pub mod parsing;

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct Symbol {
    /// Name of the attribute
    pub name: &'static str,
    /// Expected value format representation for error printing
    pub expected: &'static str,
    support: AsyncSupport,
}

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
enum AsyncSupport {
    Sync,
    Async,
    Both,
}

impl Symbol {
    pub const fn new(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Both,
        }
    }

    pub const fn new_sync(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Sync,
        }
    }

    pub const fn new_async(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Async,
        }
    }

    pub const fn test_support<const ASYNC_OR_SYNC_SWITCH: bool>(&self) -> bool {
        if ASYNC_OR_SYNC_SWITCH {
            matches!(self.support, AsyncSupport::Async | AsyncSupport::Both)
        } else {
            matches!(self.support, AsyncSupport::Sync | AsyncSupport::Both)
        }
    }
}

/// `borsh` - top level prefix in nested meta attribute
pub const BORSH: Symbol = Symbol::new("borsh", "borsh(...)");
/// `bound` - sub-borsh nested meta, field-level only; `BorshSerialize` and `BorshDeserialize` contexts
pub const BOUND: Symbol = Symbol::new_sync("bound", "bound(...)");
/// `async_bound` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
#[cfg(feature = "async")]
pub const ASYNC_BOUND: Symbol = Symbol::new_async("async_bound", "async_bound(...)");
///  `use_discriminant` - sub-borsh nested meta, item-level only, enums only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
pub const USE_DISCRIMINANT: Symbol = Symbol::new("use_discriminant", "use_discriminant = ...");
/// `serialize` - sub-bound nested meta attribute
pub const SERIALIZE: Symbol = Symbol::new("serialize", "serialize = ...");
/// `deserialize` - sub-bound nested meta attribute
pub const DESERIALIZE: Symbol = Symbol::new("deserialize", "deserialize = ...");
/// `skip` - sub-borsh nested meta, field-level only attribute;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const SKIP: Symbol = Symbol::new("skip", "skip");
/// `init` - sub-borsh nested meta, item-level only attribute; `BorshDeserialize` and `BorshDeserializeAsync` contexts
pub const INIT: Symbol = Symbol::new("init", "init = ...");
/// `serialize_with` - sub-borsh nested meta, field-level only; `BorshSerialize` context
pub const SERIALIZE_WITH: Symbol = Symbol::new_sync("serialize_with", "serialize_with = ...");
/// `deserialize_with` - sub-borsh nested meta, field-level only; `BorshDeserialize` context
pub const DESERIALIZE_WITH: Symbol = Symbol::new_sync("deserialize_with", "deserialize_with = ...");
/// `serialize_with_async` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` context
#[cfg(feature = "async")]
pub const SERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("serialize_with_async", "serialize_with_async = ...");
/// `deserialize_with_async` - sub-borsh nested meta, field-level only; `BorshDeserializeAsync` context
#[cfg(feature = "async")]
pub const DESERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("deserialize_with_async", "deserialize_with_async = ...");
/// `crate` - sub-borsh nested meta, item-level only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const CRATE: Symbol = Symbol::new("crate", "crate = ...");

#[cfg(feature = "schema")]
pub mod schema_keys {
    use super::Symbol;

    /// `schema` - sub-borsh nested meta, `BorshSchema` context
    pub const SCHEMA: Symbol = Symbol::new("schema", "schema(...)");
    /// `params` - sub-schema nested meta, field-level only attribute
    pub const PARAMS: Symbol = Symbol::new("params", "params = ...");
    /// `serialize_with` - sub-borsh nested meta, field-level only, `BorshSerialize` context
    /// `with_funcs` - sub-schema nested meta, field-level only attribute
    pub const WITH_FUNCS: Symbol = Symbol::new("with_funcs", "with_funcs(...)");
    /// `declaration` - sub-with_funcs nested meta, field-level only attribute
    pub const DECLARATION: Symbol = Symbol::new("declaration", "declaration = ...");
    /// `definitions` - sub-with_funcs nested meta, field-level only attribute
    pub const DEFINITIONS: Symbol = Symbol::new("definitions", "definitions = ...");
}

#[derive(Clone, Copy)]
pub enum BoundType {
    Serialize,
    Deserialize,
}
impl PartialEq<Symbol> for Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

impl<'a> PartialEq<Symbol> for &'a Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

fn get_one_attribute(attrs: &[Attribute]) -> syn::Result<Option<&Attribute>> {
    let mut attrs = attrs.iter().filter(|attr| attr.path() == BORSH);
    let borsh = attrs.next();
    if let Some(other_borsh) = attrs.next() {
        return Err(syn::Error::new_spanned(
            other_borsh,
            format!("multiple `{}` attributes not allowed", BORSH.name),
        ));
    }
    Ok(borsh)
}
