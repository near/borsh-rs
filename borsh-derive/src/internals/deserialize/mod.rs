use cfg_if::cfg_if;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{parse_quote, ExprPath, Generics, Ident, Path, Type};

use super::{
    attributes::{field, BoundType},
    generics,
};

pub mod enums;
pub mod structs;
pub mod unions;

struct GenericsOutput {
    overrides: Vec<syn::WherePredicate>,
    default_visitor: generics::FindTyParams,
    deserialize_visitor: generics::FindTyParams,
}

impl GenericsOutput {
    fn new(generics: &Generics) -> Self {
        Self {
            overrides: vec![],
            deserialize_visitor: generics::FindTyParams::new(generics),
            default_visitor: generics::FindTyParams::new(generics),
        }
    }

    fn extend<const IS_ASYNC: bool>(self, where_clause: &mut syn::WhereClause, cratename: &Path) {
        let de_trait: Path = if IS_ASYNC {
            parse_quote! { #cratename::de::BorshDeserializeAsync }
        } else {
            parse_quote! { #cratename::de::BorshDeserialize }
        };
        let default_trait: Path = parse_quote! { ::core::default::Default };
        let de_predicates =
            generics::compute_predicates(self.deserialize_visitor.process_for_bounds(), &de_trait);
        let default_predicates =
            generics::compute_predicates(self.default_visitor.process_for_bounds(), &default_trait);
        where_clause.predicates.extend(de_predicates);
        where_clause.predicates.extend(default_predicates);
        where_clause.predicates.extend(self.overrides);
    }
}

fn process_field<const IS_ASYNC: bool>(
    field: &syn::Field,
    cratename: &Path,
    body: &mut TokenStream2,
    generics: &mut GenericsOutput,
) -> syn::Result<()> {
    let parsed = field::Attributes::parse(&field.attrs)?;

    generics
        .overrides
        .extend(parsed.collect_bounds::<IS_ASYNC>(BoundType::Deserialize));
    let needs_bounds_derive = parsed.needs_bounds_derive::<IS_ASYNC>(BoundType::Deserialize);

    let field_name = field.ident.as_ref();
    let delta = if parsed.skip {
        if needs_bounds_derive {
            generics.default_visitor.visit_field(field);
        }
        field_default_output(field_name)
    } else {
        if needs_bounds_derive {
            generics.deserialize_visitor.visit_field(field);
        }
        field_output::<IS_ASYNC>(
            field_name,
            &field.ty,
            cratename,
            if IS_ASYNC {
                cfg_if! {
                    if #[cfg(feature = "async")] {
                        parsed.deserialize_with_async
                    } else {
                        None
                    }
                }
            } else {
                parsed.deserialize_with
            },
        )
    };
    body.extend(delta);
    Ok(())
}

/// function which computes derive output [proc_macro2::TokenStream]
/// of code, which deserializes single field
fn field_output<const IS_ASYNC: bool>(
    field_name: Option<&Ident>,
    field_type: &Type,
    cratename: &Path,
    deserialize_with: Option<ExprPath>,
) -> TokenStream2 {
    let default_path = || {
        let deserialize_trait = Ident::new(
            if IS_ASYNC {
                "BorshDeserializeAsync"
            } else {
                "BorshDeserialize"
            },
            proc_macro2::Span::call_site(),
        );
        parse_quote! { <#field_type as #cratename::#deserialize_trait>::deserialize_reader }
    };

    let path: ExprPath = deserialize_with.unwrap_or_else(default_path);
    let dot_await = IS_ASYNC.then(|| quote! { .await });

    if let Some(field_name) = field_name {
        quote! { #field_name: #path(reader)#dot_await?, }
    } else {
        quote! { #path(reader)#dot_await?, }
    }
}

/// function which computes derive output [proc_macro2::TokenStream]
/// of code, which deserializes single skipped field
fn field_default_output(field_name: Option<&Ident>) -> TokenStream2 {
    if let Some(field_name) = field_name {
        quote! { #field_name: ::core::default::Default::default(), }
    } else {
        quote! { ::core::default::Default::default(), }
    }
}
