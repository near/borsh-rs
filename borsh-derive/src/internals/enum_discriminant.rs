use std::{collections::HashMap, convert::TryFrom};

use proc_macro2::{Ident, TokenStream};
use quote::{quote, ToTokens};
use syn::{punctuated::Punctuated, token::Comma, Variant};

pub struct Discriminants(HashMap<Ident, TokenStream>);
impl Discriminants {
    /// Calculates the discriminant that will be assigned by the compiler.
    /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
    pub fn new(variants: &Punctuated<Variant, Comma>) -> Self {
        let mut map = HashMap::new();
        let mut next_discriminant_if_not_specified = quote! {0};

        for variant in variants {
            let this_discriminant = variant.discriminant.clone().map_or_else(
                || next_discriminant_if_not_specified,
                |(_, e)| e.into_token_stream(),
            );

            next_discriminant_if_not_specified = quote! { #this_discriminant + 1 };
            map.insert(variant.ident.clone(), this_discriminant);
        }

        Self(map)
    }

    pub fn get(
        &self,
        variant_ident: &Ident,
        use_discriminant: bool,
        variant_idx: usize,
    ) -> syn::Result<TokenStream> {
        let variant_idx = u8::try_from(variant_idx).map_err(|err| {
            syn::Error::new(
                variant_ident.span(),
                format!("up to 256 enum variants are supported: {}", err),
            )
        })?;
        let result = if use_discriminant {
            self.0.get(variant_ident).unwrap().clone() // discriminant value
        } else {
            quote! { #variant_idx }
        };
        Ok(result)
    }
}
