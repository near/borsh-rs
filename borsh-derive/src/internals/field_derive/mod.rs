use core::convert::TryFrom;
use proc_macro2::Span;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::ExprPath;
use syn::{Expr, Ident, Index};

pub enum FieldID {
    StructNamed(Ident),
    StructUnnamed(Index),
    EnumVariantNamed(Ident),
    EnumVariantUnnamed(Index),
}

impl FieldID {
    fn index(field_idx: usize) -> syn::Result<Index> {
        let index = u32::try_from(field_idx).map_err(|err| {
            syn::Error::new(
                Span::call_site(),
                format!("up to 2^32 fields are supported {}", err),
            )
        })?;

        Ok(Index {
            index,
            span: Span::call_site(),
        })
    }
    pub fn new_struct_index(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::StructUnnamed(index);
        Ok(result)
    }

    pub fn new_enum_index(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::EnumVariantUnnamed(index);
        Ok(result)
    }
    fn serialize_arg(&self) -> Expr {
        match self {
            Self::StructNamed(name) => syn::parse2(quote! { &self.#name }).unwrap(),
            Self::StructUnnamed(index) => syn::parse2(quote! { &self.#index }).unwrap(),
            Self::EnumVariantNamed(name) => syn::parse2(quote! { #name }).unwrap(),
            Self::EnumVariantUnnamed(index) => {
                let field_ident =
                    Ident::new(format!("id{}", index.index).as_str(), Span::mixed_site());
                syn::parse2(quote! { #field_ident }).unwrap()
            }
        }
    }

    /// function which computes derive output [proc_macro2::TokenStream]
    /// of code, which serializes single field
    pub fn serialize_output(
        &self,
        cratename: &Ident,
        serialize_with: Option<ExprPath>,
    ) -> TokenStream2 {
        let arg: Expr = self.serialize_arg();
        if let Some(func) = serialize_with {
            quote! {
                #func(#arg, writer)?;
            }
        } else {
            quote! {
                #cratename::BorshSerialize::serialize(#arg, writer)?;
            }
        }
    }

    pub fn enum_variant_header(&self, skipped: bool) -> Option<TokenStream2> {
        match self {
            Self::StructNamed(..) | Self::StructUnnamed(..) => None,
            Self::EnumVariantNamed(name) => {
                if !skipped {
                    Some(quote! { #name, })
                } else {
                    None
                }
            }
            Self::EnumVariantUnnamed(index) => {
                let field_ident = if skipped {
                    Ident::new(format!("_id{}", index.index).as_str(), Span::mixed_site())
                } else {
                    Ident::new(format!("id{}", index.index).as_str(), Span::mixed_site())
                };
                Some(quote! { #field_ident, })
            }
        }
    }
}
