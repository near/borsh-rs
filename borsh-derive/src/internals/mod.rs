pub mod attributes;
pub mod deserialize;
mod enum_discriminant;
mod generics;
#[cfg(feature = "schema")]
pub mod schema;
pub mod serialize;

pub mod cratename;

#[cfg(test)]
mod test_helpers;
