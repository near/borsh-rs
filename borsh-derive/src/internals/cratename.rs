use proc_macro2::Span;
use proc_macro_crate::{crate_name, FoundCrate};
use syn::{Attribute, Error, Ident, Path};

use super::attributes::item;

pub(crate) const BORSH: &str = "borsh";

pub(crate) fn get(attrs: &[Attribute]) -> Result<Path, Error> {
    let path = item::get_crate(attrs)?;
    match path {
        Some(path) => Ok(path),
        None => {
            let ident = get_from_cargo();
            Ok(ident.into())
        }
    }
}

pub(crate) fn get_from_cargo() -> Ident {
    let name = &crate_name(BORSH).unwrap();
    let name = match name {
        FoundCrate::Itself => BORSH,
        FoundCrate::Name(name) => name.as_str(),
    };
    Ident::new(name, Span::call_site())
}
