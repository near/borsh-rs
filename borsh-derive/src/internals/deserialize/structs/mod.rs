use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{Fields, ItemStruct, Path, Token};

use crate::internals::{attributes::item, deserialize, generics};

pub fn process<const IS_ASYNC: bool>(
    input: ItemStruct,
    cratename: Path,
) -> syn::Result<TokenStream2> {
    let name = &input.ident;
    let generics = generics::without_defaults(input.generics);
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
    let mut where_clause = generics::default_where(where_clause);
    let mut body = TokenStream2::new();
    let mut generics_output = deserialize::GenericsOutput::new(&generics);

    let return_value = match &input.fields {
        Fields::Named(fields) => {
            for field in &fields.named {
                deserialize::process_field::<IS_ASYNC>(
                    field,
                    &cratename,
                    &mut body,
                    &mut generics_output,
                )?;
            }
            quote! { Self { #body } }
        }
        Fields::Unnamed(fields) => {
            for field in fields.unnamed.iter() {
                deserialize::process_field::<IS_ASYNC>(
                    field,
                    &cratename,
                    &mut body,
                    &mut generics_output,
                )?;
            }
            quote! { Self( #body ) }
        }
        Fields::Unit => quote! { Self {} },
    };
    generics_output.extend::<IS_ASYNC>(&mut where_clause, &cratename);

    let deserialize_trait = Ident::new(
        if IS_ASYNC {
            "BorshDeserializeAsync"
        } else {
            "BorshDeserialize"
        },
        Span::call_site(),
    );
    let read_trait_path = if IS_ASYNC {
        quote! { async_io::AsyncRead }
    } else {
        quote! { io::Read }
    };
    let r#async = IS_ASYNC.then(|| Token![async](Span::call_site()));

    let body = if let Some(method_ident) = item::contains_initialize_with(&input.attrs)? {
        quote! {
            let mut return_value = #return_value;
            return_value.#method_ident();
            ::core::result::Result::Ok(return_value)
        }
    } else {
        quote! { ::core::result::Result::Ok(#return_value) }
    };

    Ok(quote! {
        impl #impl_generics #cratename::de::#deserialize_trait for #name #ty_generics #where_clause {
            #r#async fn deserialize_reader<__R: #cratename::#read_trait_path>(reader: &mut __R) -> ::core::result::Result<Self, #cratename::io::Error> {
                #body
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::*;
    use crate::internals::test_helpers::{
        default_cratename, local_insta_assert_snapshot, pretty_print_syn_str,
    };

    #[test]
    fn simple_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct A {
                x: u64,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_struct_with_custom_crate() {
        let item_struct: ItemStruct = parse_quote! {
            struct A {
                x: u64,
                y: String,
            }
        };

        let crate_: Path = parse_quote! { reexporter::borsh };

        let actual = process::<false>(item_struct.clone(), crate_.clone()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, crate_).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_generics() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K, V> {
                x: HashMap<K, V>,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_generic_tuple_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct TupleA<T>(T, u32);
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn bound_generics() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Key, V> where V: Value {
                x: HashMap<K, V>,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn recursive_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct CRecC {
                a: String,
                b: HashMap<String, CRecC>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_tuple_struct_borsh_skip1() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> (
                #[borsh(skip)]
                HashMap<K, V>,
                U,
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_tuple_struct_borsh_skip2() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> (
                HashMap<K, V>,
                #[borsh(skip)]
                U,
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_named_fields_struct_borsh_skip() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> {
                #[borsh(skip)]
                x: HashMap<K, V>,
                y: U,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_deserialize_bound() {
        let item_struct: ItemStruct = parse_quote! {
            struct C<T: Debug, U> {
                a: String,
                #[borsh(bound(deserialize =
                    "T: PartialOrd + Hash + Eq + borsh::de::BorshDeserialize,
                     U: borsh::de::BorshDeserialize"
                ))]
                b: HashMap<T, U>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn generic_deserialize_async_bound() {
        let item_struct: ItemStruct = parse_quote! {
            struct C<T: Debug, U> {
                a: String,
                #[borsh(async_bound(deserialize =
                    "T: PartialOrd + Hash + Eq + borsh::de::BorshDeserializeAsync,
                     U: borsh::de::BorshDeserializeAsync"
                ))]
                b: HashMap<T, U>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn test_override_automatically_added_default_trait() {
        let item_struct: ItemStruct = parse_quote! {
              struct G1<K, V, U>(
                #[borsh(skip,bound(deserialize = ""))]
                HashMap<K, V>,
                U
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn test_override_automatically_added_default_trait_async() {
        let item_struct: ItemStruct = parse_quote! {
              struct G1<K, V, U>(
                #[borsh(skip,async_bound(deserialize = ""))]
                HashMap<K, V>,
                U
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn check_deserialize_with_attr() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(deserialize_with = "third_party_impl::deserialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn check_deserialize_with_async_attr() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(deserialize_with_async = "third_party_impl::deserialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn borsh_init_func() {
        let item_struct: ItemStruct = parse_quote! {
            #[borsh(init=initialization_method)]
            struct A {
                x: u64,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }
}
