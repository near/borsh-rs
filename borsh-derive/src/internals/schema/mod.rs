use std::collections::HashSet;

use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{
    punctuated::Punctuated, token::Comma, Field, Fields, GenericParam, Generics, Ident, Path, Type,
    WherePredicate,
};

use crate::internals::{attributes::field, generics};

pub mod enums;
pub mod structs;

struct GenericsOutput {
    params_visitor: generics::FindTyParams,
}

impl GenericsOutput {
    fn new(generics: &Generics) -> Self {
        Self {
            params_visitor: generics::FindTyParams::new(generics),
        }
    }
    fn result(self, item_name: &str, cratename: &Path) -> (Vec<WherePredicate>, TokenStream2) {
        let trait_path: Path = syn::parse2(quote! { #cratename::BorshSchema }).unwrap();
        let predicates = generics::compute_predicates(
            self.params_visitor.clone().process_for_bounds(),
            &trait_path,
        );
        // Generate function that returns the name of the type.
        let declaration = declaration(
            item_name,
            cratename.clone(),
            self.params_visitor.process_for_bounds(),
        );

        (predicates, declaration)
    }
}

fn declaration(ident_str: &str, cratename: Path, params_for_bounds: Vec<Type>) -> TokenStream2 {
    // Generate function that returns the name of the type.
    let mut declaration_params = vec![];
    for type_param in params_for_bounds {
        declaration_params.push(quote! {
            <#type_param as #cratename::BorshSchema>::declaration()
        });
    }
    if declaration_params.is_empty() {
        quote! {
                #ident_str.to_string()
        }
    } else {
        quote! {
                let params = #cratename::__private::maybestd::vec![#(#declaration_params),*];
                format!(r#"{}<{}>"#, #ident_str, params.join(", "))
        }
    }
}

fn filter_used_params(generics: &Generics, not_skipped_type_params: HashSet<Ident>) -> Generics {
    let new_params = generics
        .params
        .clone()
        .into_iter()
        .filter(|param| match param {
            GenericParam::Lifetime(..) | GenericParam::Const(..) => true,
            GenericParam::Type(ty_param) => not_skipped_type_params.contains(&ty_param.ident),
        })
        .collect();

    let mut where_clause = generics.where_clause.clone();
    where_clause = where_clause.map(|mut clause| {
        let new_predicates: Punctuated<WherePredicate, Comma> = clause
            .predicates
            .iter()
            .filter(|predicate| match predicate {
                WherePredicate::Lifetime(..) => true,
                WherePredicate::Type(predicate_type) => generics::type_contains_some_param(
                    &predicate_type.bounded_ty,
                    &not_skipped_type_params,
                ),
                #[cfg_attr(
                    feature = "force_exhaustive_checks",
                    deny(non_exhaustive_omitted_patterns)
                )]
                _ => true,
            })
            .cloned()
            .collect();
        clause.predicates = new_predicates;
        clause
    });
    Generics {
        params: new_params,
        where_clause,
        ..generics.clone()
    }
}

fn visit_field(field: &Field, visitor: &mut generics::FindTyParams) -> syn::Result<()> {
    let parsed = field::Attributes::parse(&field.attrs)?;
    let needs_schema_params_derive = parsed.needs_schema_params_derive();
    let schema_attrs = parsed.schema;
    if !parsed.skip {
        if needs_schema_params_derive {
            visitor.visit_field(field);
        }
        // there's no need to override params when field is skipped, because when field is skipped
        // derive for it doesn't attempt to add any bounds, unlike `BorshDeserialize`, which
        // adds `Default` bound on any type parameters in skipped field

        if let Some(schema_attrs) = schema_attrs {
            if let Some(schema_params) = schema_attrs.params {
                for field::schema::ParameterOverride {
                    order_param,
                    override_type,
                    ..
                } in schema_params
                {
                    visitor.param_associated_type_insert(order_param, override_type);
                }
            }
        }
    }
    Ok(())
}

/// check param usage in fields with respect to `borsh(skip)` attribute usage
fn visit_struct_fields(fields: &Fields, visitor: &mut generics::FindTyParams) -> syn::Result<()> {
    match &fields {
        Fields::Named(fields) => {
            for field in &fields.named {
                visit_field(field, visitor)?;
            }
        }
        Fields::Unnamed(fields) => {
            for field in &fields.unnamed {
                visit_field(field, visitor)?;
            }
        }
        Fields::Unit => {}
    }
    Ok(())
}

/// check param usage in fields
fn visit_struct_fields_unconditional(fields: &Fields, visitor: &mut generics::FindTyParams) {
    match &fields {
        Fields::Named(fields) => {
            for field in &fields.named {
                visitor.visit_field(field);
            }
        }
        Fields::Unnamed(fields) => {
            for field in &fields.unnamed {
                visitor.visit_field(field);
            }
        }
        Fields::Unit => {}
    }
}
