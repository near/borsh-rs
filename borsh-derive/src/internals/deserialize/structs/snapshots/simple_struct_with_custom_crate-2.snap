---
source: borsh-derive/src/internals/deserialize/structs/mod.rs
expression: pretty_print_syn_str(actual).unwrap()
---
impl reexporter::borsh::de::BorshDeserializeAsync for A {
    async fn deserialize_reader<__R: reexporter::borsh::async_io::AsyncRead>(
        reader: &mut __R,
    ) -> ::core::result::Result<Self, reexporter::borsh::io::Error> {
        ::core::result::Result::Ok(Self {
            x: <u64 as reexporter::borsh::BorshDeserializeAsync>::deserialize_reader(
                    reader,
                )
                .await?,
            y: <String as reexporter::borsh::BorshDeserializeAsync>::deserialize_reader(
                    reader,
                )
                .await?,
        })
    }
}
